File: src/spec-sections.ts
/** Extract a top-level `## Heading` section (content after heading, before next `##`). */
export function extractSpecSection(
  body: string,
  heading: string,
): string | null {
  const escaped = heading.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const start = body.match(new RegExp(`^${escaped}\\s*$`, "m"));
  if (!start || start.index === undefined) return null;

  const after = body.slice(start.index + start[0].length);
  const next = after.match(/^## /m);
  const section = next?.index !== undefined ? after.slice(0, next.index) : after;
  const trimmed = section.trim();
  return trimmed.length > 0 ? trimmed : null;
}

/** Remove a top-level `## Heading` section from the spec body. */
export function stripSpecSection(body: string, heading: string): string {
  const escaped = heading.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const pattern = new RegExp(
    `^${escaped}\\s*$[\\s\\S]*?(?=^## |\\z)`,
    "m",
  );
  return body.replace(pattern, "").replace(/\n{3,}/g, "\n\n").trim();
}

File: python/src/curious/spec_sections.py
from __future__ import annotations

import re


def extract_spec_section(body: str, heading: str) -> str | None:
    escaped = re.escape(heading)
    start = re.search(rf"^{escaped}\s*$", body, re.M)
    if not start:
        return None
    after = body[start.end() :]
    nxt = re.search(r"^## ", after, re.M)
    section = after[: nxt.start()] if nxt else after
    trimmed = section.strip()
    return trimmed if trimmed else None


def strip_spec_section(body: str, heading: str) -> str:
    escaped = re.escape(heading)
    pattern = rf"^{escaped}\s*$[\s\S]*?(?=^## |\Z)"
    out = re.sub(pattern, "", body, flags=re.M)
    return re.sub(r"\n{3,}", "\n\n", out).strip()
